### PR TITLE
Refactor Os::Queue into CMake selection

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

This refactors Os::Queue to follow the new OSAL interface as documented in #2725.   As part of this effort the `PriorityQueue` implementation as included.  An initial pass at condition variables was implemented as these were needed for the `PriorityQueue` implementation.

### CODE REVIEW

File: Os/Queue.hpp [hunk 1]
```diff
@@ -1,94 +1,312 @@
...
-            //! \return queue creation status
-            QueueStatus createInternal(const Fw::StringBase &name, NATIVE_INT_TYPE depth, NATIVE_INT_TYPE msgSize); //!<  create a message queue
-            POINTER_CAST m_handle; //!<  handle for implementation specific queue
-            QueueString m_name; //!< queue name
-#if FW_QUEUE_REGISTRATION
-            static QueueRegistry* s_queueRegistry; //!< pointer to registry
-#endif
-            static NATIVE_INT_TYPE s_numQueues; //!< tracks number of queues in the system
+//! \brief QueueHandle parent class
+class QueueHandle {};
```
Comment (thomas-bc) [inline Os/Queue.hpp:19]: add getHandle to interface

File: Os/Queue.hpp [hunk 2]
```diff
@@ -1,94 +1,312 @@
...
+    const QueueString& getName() const;
+
+    //! \brief get number of queues system-wide
+    static FwSizeType getNumQueues();
+
+  private:
+    QueueString m_name;                           //!< queue name
+    FwSizeType m_depth;                           //!< Queue depth
+    FwSizeType m_size;                            //!< Maximum message size
+    static std::atomic<FwSizeType> s_queueCount;  //!< Count of the number of queues
```
Comment (thomas-bc) [inline Os/Queue.hpp]: use mutex instead of std::atomic

File: Os/Queue.hpp [hunk 3]
```diff
@@ -1,94 +1,312 @@
...
+    enum Status {
+        OP_OK,             //!<  message sent/received okay
+        ALREADY_CREATED,   //!<  creating an already created queue
+        EMPTY,             //!<  If non-blocking, all the messages have been drained.
+        UNINITIALIZED,     //!<  Queue wasn't initialized successfully
+        SIZE_MISMATCH,     //!<  attempted to send or receive with buffer too large, too small
+        SEND_ERROR,        //!<  message send error
+        RECEIVE_ERROR,     //!<  message receive error
+        INVALID_PRIORITY,  //!<  invalid priority requested
+        EMPTY_BUFFER,      //!<  supplied buffer is empty
```
Comment (thomas-bc) [inline Os/Queue.hpp]: empty buffer should not be an error

File: Os/Generic/PriorityQueue.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+#include <cstring>
+#include <mutex>
+#include <new>
+
+namespace Os {
+namespace Generic {
+
+FwSizeType PriorityQueueHandle ::find_index() {
+    // This function is predicated on the definition of unsigned "overflow"
+    static_assert(not std::numeric_limits<FwSizeType>::is_signed, "FwSizeType is presumed to be unsigned");
```
Comment (thomas-bc) [inline Os/Generic/PriorityQueue.cpp]: Ensure `static_assert()` is entirely ran at compile-time and is scrubbed from compiled code

File: Os/Generic/PriorityQueue.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    // This function will never be called when full, this assert ensures that the start index never surpasses the end
+    // index.
+    FW_ASSERT(this->m_startIndex != this->m_stopIndex);
+    FwSizeType index = this->m_indices[this->m_startIndex % this->m_depth];
```
Comment (thomas-bc) [inline Os/Generic/PriorityQueue.cpp:15]: Fix the wraparound bug

Comment (thomas-bc) [reply Os/Generic/PriorityQueue.cpp:15]: Tim asks for a section in the SDD about the math of it all (ideally diagrams?)

File: Os/Generic/PriorityQueue.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+PriorityQueue::~PriorityQueue() {
+    delete[] this->m_handle.m_data;
+    delete[] this->m_handle.m_indices;
+    delete[] this->m_handle.m_sizes;
+}
+
+QueueInterface::Status PriorityQueue::create(const Fw::StringBase& name, FwSizeType depth, FwSizeType messageSize) {
+    // Allocate indices list
+    FwSizeType* indices = new (std::nothrow) FwSizeType[depth];
```
Comment (thomas-bc) [inline Os/Generic/PriorityQueue.cpp:54]: assert that it's not created already

File: Os/Generic/PriorityQueue.cpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            return QueueInterface::Status::FULL;
+        }
+        // Will loop and block until full is false
+        while (this->m_handle.m_heap.isFull()) {
+            this->m_handle.m_full.wait(this->m_handle.m_data_lock);
+        }
+        FwSizeType index = this->m_handle.find_index();
+
+        bool pushed = this->m_handle.m_heap.push(priority, index);
+        if (not pushed) {
```
Comment (thomas-bc) [inline Os/Generic/PriorityQueue.cpp]: may want to assert these conditions instead cause we are protected by the scope lock

**Review by thomas-bc — COMMENTED** [review]
Peer review comments

File: Os/Generic/PriorityQueue.cpp (hunk 4 above)
Comment (LeStarch) [reply Os/Generic/PriorityQueue.cpp]: Confirmed "If bool-constexpr is well-formed and evaluates to true, or is evaluated in the context of a template definition and the template is uninstantiated, **this declaration has no effect.**"

File: Os/Generic/PriorityQueue.cpp (hunk 5 above)
Comment (LeStarch) [reply Os/Generic/PriorityQueue.cpp:15]: Fixed part one!

**Comment by LeStarch** [comment]
Add Os::Model for Queue status.

File: Os/Generic/PriorityQueue.cpp (hunk 5 above)
Comment (LeStarch) [reply Os/Generic/PriorityQueue.cpp:15]: Confirmed the SDD was a request for documentation and further review.  Will track as an issue.

**Review by thomas-bc — APPROVED** [review]

**Pull request merged**